File: platform/cli/src/llm.ts
import type { ChangeType } from './workspace';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LLMResult {
  title: string;        // "feat(auth): add login API endpoints"
  description: string;  // "- Add /api/auth/login endpoint\n- JWT token generation"
}

export interface LLMProvider {
  name: string;
  generate(prompt: string): Promise<string>;
}

// ---------------------------------------------------------------------------
// Provider detection (in order)
// 1. DEVPOD_LLM_URL (custom OpenAI-compatible endpoint)
// 2. ANTHROPIC_API_KEY
// 3. OPENAI_API_KEY
// 4. Check Ollama at localhost:11434
// 5. Fallback (no LLM)
// ---------------------------------------------------------------------------

const LLM_TIMEOUT = 5000;

export function detectProvider(): LLMProvider | null {
  // 1. Custom endpoint
  if (process.env.DEVPOD_LLM_URL) {
    return createOpenAIProvider(
      'custom',
      process.env.DEVPOD_LLM_URL,
      process.env.DEVPOD_LLM_KEY || '',
      process.env.DEVPOD_LLM_MODEL || 'default',
    );
  }

  // 2. Anthropic
  if (process.env.ANTHROPIC_API_KEY) {
    return createAnthropicProvider(
      process.env.ANTHROPIC_API_KEY,
      process.env.DEVPOD_LLM_MODEL || 'claude-sonnet-4-20250514',
    );
  }

  // 3. OpenAI
  if (process.env.OPENAI_API_KEY) {
    return createOpenAIProvider(
      'openai',
      'https://api.openai.com',
      process.env.OPENAI_API_KEY,
      process.env.DEVPOD_LLM_MODEL || 'gpt-4o-mini',
    );
  }

  // 4. Ollama — we cannot do a synchronous check here, so we create
  //    the provider optimistically. If Ollama is down, the call will
  //    fail within the timeout and the caller falls back.
  return createOpenAIProvider(
    'ollama',
    'http://localhost:11434',
    '',
    process.env.DEVPOD_LLM_MODEL || 'llama3.2',
  );
}

export function isLLMAvailable(): boolean {
  return !!(
    process.env.DEVPOD_LLM_URL ||
    process.env.ANTHROPIC_API_KEY ||
    process.env.OPENAI_API_KEY
  );
  // Ollama availability requires a network check and is best-effort,
  // so we don't include it in this synchronous check.
}

// ---------------------------------------------------------------------------
// Anthropic provider
// ---------------------------------------------------------------------------

function createAnthropicProvider(apiKey: string, model: string): LLMProvider {
  return {
    name: 'anthropic',
    async generate(prompt: string): Promise<string> {
      const response = await fetchWithTimeout(
        'https://api.anthropic.com/v1/messages',
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'x-api-key': apiKey,
            'anthropic-version': '2023-06-01',
          },
          body: JSON.stringify({
            model,
            max_tokens: 300,
            messages: [{ role: 'user', content: prompt }],
          }),
        },
        LLM_TIMEOUT,
      );
      const json = (await response.json()) as { content?: { text?: string }[] };
      return json.content?.[0]?.text || '';
    },
  };
}

// ---------------------------------------------------------------------------
// OpenAI-compatible provider (also used for Ollama and custom endpoints)
// ---------------------------------------------------------------------------

function createOpenAIProvider(
  name: string,
  baseUrl: string,
  apiKey: string,
  model: string,
): LLMProvider {
  return {
    name,
    async generate(prompt: string): Promise<string> {
      const url = `${baseUrl.replace(/\/$/, '')}/v1/chat/completions`;
      const headers: Record<string, string> = { 'Content-Type': 'application/json' };
      if (apiKey) {
        headers['Authorization'] = `Bearer ${apiKey}`;
      }

      const response = await fetchWithTimeout(
        url,
        {
          method: 'POST',
          headers,
          body: JSON.stringify({
            model,
            messages: [
              { role: 'system', content: 'You generate concise commit messages.' },
              { role: 'user', content: prompt },
            ],
            max_tokens: 300,
            temperature: 0.3,
          }),
        },
        LLM_TIMEOUT,
      );
      const json = (await response.json()) as { choices?: { message?: { content?: string } }[] };
      return json.choices?.[0]?.message?.content || '';
    },
  };
}

// ---------------------------------------------------------------------------
// HTTP helper with timeout
// ---------------------------------------------------------------------------

async function fetchWithTimeout(url: string, init: RequestInit, timeoutMs: number): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...init, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

// ---------------------------------------------------------------------------
// Prompt templates
// ---------------------------------------------------------------------------

function commitMessagePrompt(diff: string, type: ChangeType, scope: string): string {
  return `Generate a conventional commit message for this diff.

Type: ${type}
Scope: ${scope}

Format:
- Title: type(scope): description (max 72 chars, lowercase)
- Description: 2-3 bullet points starting with "- "

Rules:
- Be specific about what was added/changed/removed
- No AI attribution
- No generic messages like "update files"

Diff:
${diff.slice(0, 3000)}`;
}

function typeDetectionPrompt(diff: string): string {
  return `Classify this code change as one of: feature, fix, docs, chore

- feature: new functionality or capability
- fix: bug fix or error correction
- docs: documentation changes only
- chore: tooling, dependencies, config, refactoring

Respond with ONLY the type word, nothing else.

Diff:
${diff.slice(0, 2000)}`;
}

// ---------------------------------------------------------------------------
// Response parsing
// ---------------------------------------------------------------------------

function parseCommitMessageResponse(text: string, type: ChangeType, scope: string): LLMResult {
  // Try to extract title line (first non-empty line)
  const lines = text.trim().split('\n').filter(Boolean);
  if (lines.length === 0) {
    return { title: `${type}(${scope}): update files`, description: '' };
  }

  // Look for explicit TITLE:/DESCRIPTION: format
  const titleMatch = text.match(/TITLE:\s*(.+)/i);
  const descMatch = text.match(/DESCRIPTION:\s*([\s\S]*)/i);
  if (titleMatch) {
    return {
      title: titleMatch[1]!.trim().slice(0, 72),
      description: descMatch ? descMatch[1]!.trim() : '',
    };
  }

  // Otherwise take first line as title, rest as description
  let title = lines[0]!.trim();
  // Ensure title follows conventional commit format
  if (!title.match(/^\w+\(/)) {
    title = `${type}(${scope}): ${title}`;
  }

  const descLines = lines.slice(1).filter((l) => l.startsWith('- '));
  return {
    title: title.slice(0, 72),
    description: descLines.join('\n'),
  };
}

function parseTypeResponse(text: string): ChangeType {
  const cleaned = text.toLowerCase().trim();
  if (cleaned.includes('fix')) return 'fix';
  if (cleaned.includes('docs') || cleaned.includes('documentation')) return 'docs';
  if (cleaned.includes('chore')) return 'chore';
  if (cleaned.includes('feature') || cleaned.includes('feat')) return 'feature';
  return 'unknown';
}

// ---------------------------------------------------------------------------
// Main API
// ---------------------------------------------------------------------------

export async function generateDiffMessage(diff: string, type: ChangeType, scope: string): Promise<LLMResult> {
  const provider = detectProvider();
  if (!provider) {
    return { title: `${type}(${scope}): update files`, description: '' };
  }

  try {
    const prompt = commitMessagePrompt(diff, type, scope);
    const text = await provider.generate(prompt);
    if (!text) {
      return { title: `${type}(${scope}): update files`, description: '' };
    }
    return parseCommitMessageResponse(text, type, scope);
  } catch {
    return { title: `${type}(${scope}): update files`, description: '' };
  }
}

export async function detectChangeType(diff: string): Promise<ChangeType> {
  const provider = detectProvider();
  if (!provider) return 'unknown';

  try {
    const prompt = typeDetectionPrompt(diff);
    const text = await provider.generate(prompt);
    if (!text) return 'unknown';
    return parseTypeResponse(text);
  } catch {
    return 'unknown';
  }
}

export function generateFallbackMessage(files: string[], type: ChangeType, scope: string): LLMResult {
  if (files.length === 0) {
    return { title: `${type}(${scope}): update files`, description: '' };
  }

  const areas = new Set<string>();
  for (const f of files) {
    const parts = f.split('/');
    if (parts.length > 1) areas.add(parts[parts.length - 2]!);
    else areas.add(f);
  }
  const areaStr = Array.from(areas).slice(0, 3).join(', ');
  return {
    title: `${type}(${scope}): update ${areaStr}`.slice(0, 72),
    description: files.map((f) => `- ${f}`).slice(0, 5).join('\n'),
  };
}

File: platform/cli-go/internal/llm/llm.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

const llmTimeout = 5 * time.Second

// Result holds a generated commit title and description.
type Result struct {
	Title       string
	Description string
}

// Provider is the interface for LLM backends.
type Provider interface {
	Name() string
	Generate(prompt string) (string, error)
}

// ---------------------------------------------------------------------------
// Provider detection (in order)
// 1. DEVPOD_LLM_URL (custom OpenAI-compatible endpoint)
// 2. ANTHROPIC_API_KEY
// 3. OPENAI_API_KEY
// 4. Ollama at localhost:11434
// 5. nil (fallback)
// ---------------------------------------------------------------------------

// DetectProvider finds the best available LLM provider.
func DetectProvider() Provider {
	// 1. Custom endpoint
	if url := os.Getenv("DEVPOD_LLM_URL"); url != "" {
		key := os.Getenv("DEVPOD_LLM_KEY")
		model := os.Getenv("DEVPOD_LLM_MODEL")
		if model == "" {
			model = "default"
		}
		return newOpenAIProvider("custom", url, key, model)
	}

	// 2. Anthropic
	if key := os.Getenv("ANTHROPIC_API_KEY"); key != "" {
		model := os.Getenv("DEVPOD_LLM_MODEL")
		if model == "" {
			model = "claude-sonnet-4-20250514"
		}
		return newAnthropicProvider(key, model)
	}

	// 3. OpenAI
	if key := os.Getenv("OPENAI_API_KEY"); key != "" {
		model := os.Getenv("DEVPOD_LLM_MODEL")
		if model == "" {
			model = "gpt-4o-mini"
		}
		return newOpenAIProvider("openai", "https://api.openai.com", key, model)
	}

	// 4. Ollama — optimistic; if down, calls fail within timeout
	model := os.Getenv("DEVPOD_LLM_MODEL")
	if model == "" {
		model = "llama3.2"
	}
	return newOpenAIProvider("ollama", "http://localhost:11434", "", model)
}

// IsAvailable checks if any LLM provider is configured via environment.
// Ollama availability requires a network check, so it is excluded here.
func IsAvailable() bool {
	return os.Getenv("DEVPOD_LLM_URL") != "" ||
		os.Getenv("ANTHROPIC_API_KEY") != "" ||
		os.Getenv("OPENAI_API_KEY") != ""
}

// ---------------------------------------------------------------------------
// Anthropic provider
// ---------------------------------------------------------------------------

type anthropicProvider struct {
	apiKey string
	model  string
}

func newAnthropicProvider(apiKey, model string) Provider {
	return &anthropicProvider{apiKey: apiKey, model: model}
}

func (p *anthropicProvider) Name() string { return "anthropic" }

func (p *anthropicProvider) Generate(prompt string) (string, error) {
	body := map[string]interface{}{
		"model":      p.model,
		"max_tokens": 300,
		"messages":   []map[string]string{{"role": "user", "content": prompt}},
	}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return "", err
	}

	ctx, cancel := context.WithTimeout(context.Background(), llmTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.anthropic.com/v1/messages", bytes.NewReader(jsonBody))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", p.apiKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var result struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return "", err
	}
	if len(result.Content) > 0 {
		return result.Content[0].Text, nil
	}
	return "", nil
}

// ---------------------------------------------------------------------------
// OpenAI-compatible provider (also used for Ollama and custom endpoints)
// ---------------------------------------------------------------------------

type openAIProvider struct {
	name    string
	baseURL string
	apiKey  string
	model   string
}

func newOpenAIProvider(name, baseURL, apiKey, model string) Provider {
	return &openAIProvider{name: name, baseURL: baseURL, apiKey: apiKey, model: model}
}

func (p *openAIProvider) Name() string { return p.name }

func (p *openAIProvider) Generate(prompt string) (string, error) {
	url := strings.TrimRight(p.baseURL, "/") + "/v1/chat/completions"

	body := map[string]interface{}{
		"model": p.model,
		"messages": []map[string]string{
			{"role": "system", "content": "You generate concise commit messages."},
			{"role": "user", "content": prompt},
		},
		"max_tokens":  300,
		"temperature": 0.3,
	}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return "", err
	}

	ctx, cancel := context.WithTimeout(context.Background(), llmTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(jsonBody))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	if p.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+p.apiKey)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return "", err
	}
	if len(result.Choices) > 0 {
		return result.Choices[0].Message.Content, nil
	}
	return "", nil
}

// ---------------------------------------------------------------------------
// Prompt templates
// ---------------------------------------------------------------------------

func commitMessagePrompt(diff, changeType, scope string) string {
	truncated := diff
	if len(truncated) > 3000 {
		truncated = truncated[:3000]
	}
	return fmt.Sprintf(`Generate a conventional commit message for this diff.

Type: %s
Scope: %s

Format:
- Title: type(scope): description (max 72 chars, lowercase)
- Description: 2-3 bullet points starting with "- "

Rules:
- Be specific about what was added/changed/removed
- No AI attribution
- No generic messages like "update files"

Diff:
%s`, changeType, scope, truncated)
}

func typeDetectionPrompt(diff string) string {
	truncated := diff
	if len(truncated) > 2000 {
		truncated = truncated[:2000]
	}
	return fmt.Sprintf(`Classify this code change as one of: feature, fix, docs, chore

- feature: new functionality or capability
- fix: bug fix or error correction
- docs: documentation changes only
- chore: tooling, dependencies, config, refactoring

Respond with ONLY the type word, nothing else.

Diff:
%s`, truncated)
}

// ---------------------------------------------------------------------------
// Response parsing
// ---------------------------------------------------------------------------

func parseCommitMessageResponse(text, changeType, scope string) Result {
	lines := splitNonEmpty(strings.TrimSpace(text))
	if len(lines) == 0 {
		return Result{Title: fmt.Sprintf("%s(%s): update files", changeType, scope)}
	}

	// Look for explicit TITLE:/DESCRIPTION: format
	titleRe := regexp.MustCompile(`(?i)TITLE:\s*(.+)`)
	descRe := regexp.MustCompile(`(?is)DESCRIPTION:\s*(.*)`)

	if match := titleRe.FindStringSubmatch(text); match != nil {
		title := strings.TrimSpace(match[1])
		if len(title) > 72 {
			title = title[:72]
		}
		desc := ""
		if dmatch := descRe.FindStringSubmatch(text); dmatch != nil {
			desc = strings.TrimSpace(dmatch[1])
		}
		return Result{Title: title, Description: desc}
	}

	// Otherwise take first line as title, rest as description
	title := strings.TrimSpace(lines[0])
	conventionalRe := regexp.MustCompile(`^\w+\(`)
	if !conventionalRe.MatchString(title) {
		title = fmt.Sprintf("%s(%s): %s", changeType, scope, title)
	}
	if len(title) > 72 {
		title = title[:72]
	}

	var descLines []string
	for _, l := range lines[1:] {
		if strings.HasPrefix(l, "- ") {
			descLines = append(descLines, l)
		}
	}
	return Result{Title: title, Description: strings.Join(descLines, "\n")}
}

func parseTypeResponse(text string) string {
	cleaned := strings.ToLower(strings.TrimSpace(text))
	if strings.Contains(cleaned, "fix") {
		return "fix"
	}
	if strings.Contains(cleaned, "docs") || strings.Contains(cleaned, "documentation") {
		return "docs"
	}
	if strings.Contains(cleaned, "chore") {
		return "chore"
	}
	if strings.Contains(cleaned, "feature") || strings.Contains(cleaned, "feat") {
		return "feature"
	}
	return "unknown"
}

func splitNonEmpty(s string) []string {
	var result []string
	for _, line := range strings.Split(s, "\n") {
		if strings.TrimSpace(line) != "" {
			result = append(result, line)
		}
	}
	return result
}

// ---------------------------------------------------------------------------
// Main API
// ---------------------------------------------------------------------------

// GenerateDiffMessage uses LLM to create a conventional commit message.
func GenerateDiffMessage(diff, changeType, scope string) (Result, error) {
	provider := DetectProvider()
	if provider == nil {
		return Result{Title: fmt.Sprintf("%s(%s): update files", changeType, scope)}, nil
	}

	prompt := commitMessagePrompt(diff, changeType, scope)
	text, err := provider.Generate(prompt)
	if err != nil || text == "" {
		return Result{Title: fmt.Sprintf("%s(%s): update files", changeType, scope)}, nil
	}
	return parseCommitMessageResponse(text, changeType, scope), nil
}

// DetectChangeType uses LLM to classify the change.
func DetectChangeType(diff string) (string, error) {
	provider := DetectProvider()
	if provider == nil {
		return "unknown", nil
	}

	prompt := typeDetectionPrompt(diff)
	text, err := provider.Generate(prompt)
	if err != nil || text == "" {
		return "unknown", nil
	}
	return parseTypeResponse(text), nil
}

// GenerateFallbackMessage creates a message from file names (no LLM).
func GenerateFallbackMessage(files []string, changeType, scope string) Result {
	if len(files) == 0 {
		return Result{Title: fmt.Sprintf("%s(%s): update files", changeType, scope)}
	}

	areas := make(map[string]struct{})
	var areaOrder []string
	for _, f := range files {
		parts := strings.Split(f, "/")
		var area string
		if len(parts) > 1 {
			area = parts[len(parts)-2]
		} else {
			area = f
		}
		if _, exists := areas[area]; !exists {
			areas[area] = struct{}{}
			areaOrder = append(areaOrder, area)
		}
	}

	if len(areaOrder) > 3 {
		areaOrder = areaOrder[:3]
	}
	areaStr := strings.Join(areaOrder, ", ")

	title := fmt.Sprintf("%s(%s): update %s", changeType, scope, areaStr)
	if len(title) > 72 {
		title = title[:72]
	}

	descFiles := files
	if len(descFiles) > 5 {
		descFiles = descFiles[:5]
	}
	var descLines []string
	for _, f := range descFiles {
		descLines = append(descLines, "- "+f)
	}

	return Result{Title: title, Description: strings.Join(descLines, "\n")}
}
